src/nyx/features/power_spectrum.cpp

```diff
@@ -0,0 +1,220 @@
+    {
+        int lab = (*ptrLabels)[i];
+        LR& r = (*ptrLabelData)[lab];
+
+        PowerSpectrumFeature fsf;
+
+        fsf.calculate (r);
+
+        fsf.save_value (r.fvals);
+    }
+}
+
+void PowerSpectrumFeature::parallel_process(std::vector<int>& roi_labels, std::unordered_map <int, LR>& roiData, int n_threads)
+{
+	size_t jobSize = roi_labels.size(),
+		workPerThread = jobSize / n_threads;
+
+	runParallel(PowerSpectrumFeature::parallel_process_1_batch, n_threads, workPerThread, jobSize, &roi_labels, &roiData);
+}
+
+void PowerSpectrumFeature::parallel_process_1_batch(size_t firstitem, size_t lastitem, std::vector<int>* ptrLabels, std::unordered_map <int, LR>* ptrLabelData)
+{
+	// Calculate the feature for each batch ROI item 
+	for (auto i = firstitem; i < lastitem; i++)
+	{
+		// Get ahold of ROI's label and cache
+		int roiLabel = (*ptrLabels)[i];
+		LR& r = (*ptrLabelData)[roiLabel];
+
+		// Skip the ROI if its data is invalid to prevent nans and infs in the output
+		if (r.has_bad_data())
+			continue;
+
+		// Calculate the feature and save it in ROI's csv-friendly b uffer 'fvals'
+		PowerSpectrumFeature f;
+		f.calculate(r);
+		f.save_value(r.fvals);
+	}
+}
+
+void PowerSpectrumFeature::save_value(std::vector<std::vector<double>>& feature_vals) {
+    
+    feature_vals[(int)Feature2D::POWER_SPECTRUM_SLOPE][0] = slope_;
+
+}
+
+double PowerSpectrumFeature::power_spectrum_slope(const ImageMatrix& Im) {
+
+    readOnlyPixels image = Im.ReadablePixels();
+
+    int rows = Im.height;
+    int cols = Im.width;
+
+    std::vector<int> radii;
+    std::vector<double> magnitude, power;
+
+    std::tie(radii, magnitude, power) = rps(image, rows, cols);
+
+    std::vector<double> result;
+
+    if (std::accumulate(magnitude.begin(), magnitude.end(), 0.) > 0 && 
+        std::set<unsigned int>(image.begin(), image.end()).size() > 0) {
+
+        std::vector<double> log_radii;
+        for (int i = 0; i < magnitude.size(); ++i) {
+
+            if (magnitude[i] < 0 || !std::isfinite(std::log(power[i]))) {
+                radii.erase(radii.begin() + i);
+                power.erase(power.begin() + i);
+            } else {
+                log_radii.push_back(std::log(radii[i]));
+            }
+        }
+        
+        if (radii.size() > 1) {
+            std::vector<std::vector<double>> A (radii.size(), std::vector<double>(2, 1));
+
+            for (int i = 0; i < radii.size(); ++i) {
+                A[i][0] = std::log(radii[i]);
+            }
+
+            // apply log to power vector
+            std::transform(power.begin(), power.end(), power.begin(), [](double x) { return std::log(x); });
+            
+            return lstsq(A, power)[0]; // get slope from least squares
+        }
+    }
+
+    return 0;
+}
+
+std::vector<double> PowerSpectrumFeature::invariant(std::vector<unsigned int> image) {
+
+    
+    double mean_value = std::accumulate(image.begin(), image.end(), 0.0)/image.size();
```

Comment:
How will this always be an integer? std::accumulate returns a double because the initial value is a double and then this double is divided by an integer, which can end up as a double with a decimal

Reply:
You are right, return type comes from the 3rd argument. 
My habit to assume it to come from the container elements is outdated.